rust-runtime/smithy-json/src/serialize.rs

```diff
@@ -0,0 +1,451 @@
+        self
+    }
+
+    /// Writes the boolean `value` to the array.
+    pub fn boolean(&mut self, value: bool) -> &mut Self {
+        self.comma_delimit();
+        self.json.push_str(match value {
+            true => "true",
+            _ => "false",
+        });
+        self
+    }
+
+    /// Writes a string to the array.
+    pub fn string(&mut self, value: &str) -> &mut Self {
+        self.comma_delimit();
+        append_string(&mut self.json, value);
+        self
+    }
+
+    /// Writes a string `value` to the array without escaping it.
+    pub fn trusted_string(&mut self, value: &str) -> &mut Self {
+        self.comma_delimit();
+        append_string(&mut self.json, value);
+        self
+    }
+
+    /// Writes a number `value` to the array.
+    pub fn number(&mut self, value: Number) -> &mut Self {
+        self.comma_delimit();
+        append_number(&mut self.json, value);
+        self
+    }
+
+    /// Writes an Instant `value` with the given `key` and `format`.
+    pub fn instant(&mut self, instant: &Instant, format: Format) -> &mut Self {
+        self.comma_delimit();
+        append_instant(&mut self.json, instant, format);
+        self
+    }
+
+    /// Starts a nested array inside of the array.
+    pub fn start_array(&mut self) -> JsonArrayWriter {
+        self.comma_delimit();
+        JsonArrayWriter::new(&mut self.json)
+    }
+
+    /// Starts a nested object inside of the array.
+    pub fn start_object(&mut self) -> JsonObjectWriter {
+        self.comma_delimit();
+        JsonObjectWriter::new(&mut self.json)
+    }
+
+    /// Finishes the array.
+    pub fn finish(self) {
+        self.json.push(']');
+    }
+
+    fn comma_delimit(&mut self) {
+        if self.started {
+            self.json.push(',');
+        }
+        self.started = true;
+    }
+}
+
+fn append_string(json: &mut String, value: &str) {
+    append_trusted_string(json, &escape_string(value));
+}
+
+fn append_trusted_string(json: &mut String, value: &str) {
+    json.push('"');
+    json.push_str(value);
+    json.push('"');
+}
+
+fn append_instant(json: &mut String, value: &Instant, format: Format) {
+    let formatted = value.fmt(format);
+    match format {
+        Format::EpochSeconds => json.push_str(&formatted),
+        _ => append_string(json, &formatted),
+    }
+}
+
+const F64_EXPONENT_MASK: u64 = 0x7FF0000000000000;
+
+fn append_number(json: &mut String, value: Number) {
+    match value {
+        Number::PosInt(value) => {
+            json.push_str(itoa::Buffer::new().format(value));
+        }
+        Number::NegInt(value) => {
+            json.push_str(itoa::Buffer::new().format(value));
+        }
+        Number::Float(value) => {
+            // If the value is NaN, Infinity, or -Infinity
+            if value.to_bits() & F64_EXPONENT_MASK == F64_EXPONENT_MASK {
+                json.push_str("null");
+            } else {
+                let mut buffer = ryu::Buffer::new();
```

Comment:
I would add a comment that `ryu::Buffer::new()` and `itoa::Buffer:new()` are cheap. I assumed this allocated until I read the code, might as well save some future person from trying to optimize this.

In fact, `itoa::write` just calls `itoa::Buffer::new()`!